dallinger/docker/tools.py

```diff
@@ -189,22 +177,18 @@ def monitor(self, listener):
             if listener(line) is self.MONITOR_STOP:
                 return
 
-    def run_compose(self, compose_commands):
+    def run_compose(self, compose_commands: str):
         """Run a command in the (already built) tmp directory of the current experiment
-        `compose_commands` should be a list of tokens to be passed to docker compose.
+        `compose_commands` should be a string to be appended to the docker-compose command.
         Examples:
         # return the output of `docker-compose ps`
-        compose_commands = ["ps"]
-        # Run redis-cli ping inside the redis container and return its output
-        compose_commands = ["exec", "redis", "redis-cli", "ping"]
+        compose_commands = "ps"
+        # Run `redis-cli ping` inside the `redis` container and return its output
+        compose_commands = "exec redis redis-cli ping"]
         """
         return check_output(
-            [
-                "docker-compose",
-                "-f",
-                f"{self.tmp_dir}/docker-compose.yml",
-            ]
-            + compose_commands
+            f"docker-compose -f {self.tmp_dir}/docker-compose.yml {compose_commands}",
+            shell=True,
```

Comment:
I have some reservations about this switch to passing a space-separated string instead of a list, in combination with `shell=True`:
1.  it's a departure from how we use subprocess everywhere else
2. https://docs.python.org/3/library/subprocess.html#security-considerations
3. (related to 1) we've historically had a hard time exposing the useful content of subprocess error output into heroku logs, and quite a bit of prior art exists to address this (see `run_command` in `dallinger.utils.py`). I'm not sure that will be an issue here, but if it were, using the non `shell=True` API makes composing command calls easier

Reply:
Agreed. I didn't really think it through and was blinded by the convenience of having the sell tokenize commands.